airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py

```diff
@@ -0,0 +1,312 @@
+
+    @retry_pattern(backoff.expo, FacebookRequestError, max_tries=5, factor=5)
+    def _get_insights(self, params):
+        return self._api.account.get_insights(params=params)
+
+
+class Client(BaseClient):
+    def __init__(self, account_id: str, access_token: str, start_date: str):
+        super().__init__()
+        self._api = FacebookAdsApi.init(access_token=access_token)
+        self._account = self._find_account(account_id)
+        self._start_date = isoparse(start_date)
+
+    def stream__campaigns(self, fields):
+        yield from CampaignAPI(self).list(fields)
+
+    def stream__ad_sets(self, fields):
+        yield from AdSetsAPI(self).list(fields)
+
+    def stream__ads(self, fields):
+        yield from AdsAPI(self).list(fields)
+
+    def stream__adcreatives(self, fields):
+        yield from AdCreativeAPI(self).list(fields)
+
+    def stream__ads_insight(self, fields):
+        client = AdsInsightAPI(self, start_date=self._start_date)
+        yield from client.list(fields)
+
+    def stream__ads_insight_age_and_gender(self, fields):
+        breakdowns = ['age', 'gender']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    def stream__ads_insight_country(self, fields):
+        breakdowns = ['country']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    def stream__ads_insight_platform_and_device(self, fields):
+        breakdowns = ['publisher_platform', 'platform_position', 'impression_device']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    def stream__ads_insight_region(self, fields):
+        breakdowns = ['region']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    def stream__ads_insight_dma(self, fields):
+        breakdowns = ['dma']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    @staticmethod
+    def _find_account(account_id: str):
+        accounts = fb_user.User(fbid='me').get_ad_accounts()
+        for account in accounts:
+            if account['account_id'] == account_id:
+                return account
+
+        raise FacebookAPIException("Couldn't find account with id {}".format(account_id))
+
+    # FIXME: implement
+    def health_check(self) -> Tuple[bool, str]:
+        return True, "Error"
+
+    # FIXME: filter fields
+    @staticmethod
+    def _get_fields_from_stream(stream: AirbyteStream) -> List[str]:
+        print(stream.json_schema)
+        return []
+
+    def read_stream(self, stream: AirbyteStream) -> Generator[AirbyteRecordMessage, None, None]:
+        """Yield records from stream"""
+        method = self._stream_methods.get(stream.name)
```

Comment:
I would consider moving this to the base class.
`method = self.get_method(stream.name)` would be more concise and expressive.

Reply:
moved to base class, and fixed existing derived classes to accept `kwargs`